### PR TITLE
Workaround error: image did not contain usr/bin/oc

### PR DESCRIPTION
**Pull request by shellyyang1989**
Fixes the oc extract issue:

```
Extracting oc
error: image did not contain usr/bin/oc
Failed to extract oc binary, retry...
error: image did not contain usr/bin/oc
```

But we have injected oc by cli: latest in the toimage ref.

```
ref:
  as: cucushift-upgrade-toimage
  from: ansible
  timeout: 24h
  grace_period: 10m
  commands: cucushift-upgrade-toimage-commands.sh
  cli: latest
```

### CODE REVIEW

**Comment by shellyyang1989** [comment]
cc @liangxia for review. Thanks!

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 1]
```diff
@@ -150,7 +150,7 @@ function extract_oc(){
     echo -e "Extracting oc\n"
     local retry=5 tmp_oc="/tmp/client-2"
     mkdir -p ${tmp_oc}
-    while ! (oc adm release extract -a "${CLUSTER_PROFILE_DIR}/pull-secret" --command=oc --to=${tmp_oc} ${TARGET});
+    while ! (/cli/oc adm release extract -a "${CLUSTER_PROFILE_DIR}/pull-secret" --command=oc --to=${tmp_oc} ${TARGET});
```
Comment (liangxia) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:153]: The error is `error: image did not contain usr/bin/oc`,  not the same as `oc: command not found...`

I don't think the change fix the issue.

Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:153]: oh, do you have any idea why we ran into the error?

Comment (liangxia) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:153]: Looks like the command `oc adm release extract` try to extract oc binary from location `usr/bin/oc` in `TARGET`, but TARGET does not have oc binary in `usr/bin/oc`

Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:153]: Thanks!

Comment (shellyyang1989) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh]: I logged in to the upgrade pod and found the oc actually has been extracted. But I can't reproduce it in my end. So I'm not sure if it's a product issue or a prow dedicated issue. The change just tries to work around the issue for now.

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 2]
```diff
@@ -150,9 +150,12 @@ function extract_oc(){
...
     local retry=5 tmp_oc="/tmp/client-2"
     mkdir -p ${tmp_oc}
-    while ! (oc adm release extract -a "${CLUSTER_PROFILE_DIR}/pull-secret" --command=oc --to=${tmp_oc} ${TARGET});
+    echo "oc adm release extract -a ${CLUSTER_PROFILE_DIR}/pull-secret --command=oc --to=${tmp_oc} ${TARGET}"
+    while [ ! -x "${tmp_oc}/oc" ];
     do
-        echo >&2 "Failed to extract oc binary, retry..."
+        echo "Target oc binary doesn't exist, extracting oc..."
+        # To work around the issue error: image did not contain usr/bin/oc, we always return 0 
+        oc adm release extract -a "${CLUSTER_PROFILE_DIR}/pull-secret" --command=oc --to=${tmp_oc} ${TARGET} || true
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:158]: This means when "image did not contain usr/bin/oc", retry to extract it?

**Comment by jianlinliu** [comment]
@shellyyang1989  From the log - https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/33787/rehearse-33787-periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-nightly-4.12-upgrade-from-stable-4.4-aws-ipi/1590266972224884736, "Error: unknown flag: --allow-explicit-upgrade", seem like another bug.

But the workaround in the PR looks working well.

**Comment by shellyyang1989** [comment]
> @shellyyang1989 From the log - https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/33787/rehearse-33787-periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-nightly-4.12-upgrade-from-stable-4.4-aws-ipi/1590266972224884736, "Error: unknown flag: --allow-explicit-upgrade", seem like another bug.
> 
> But the workaround in the PR looks working well.

Right. But the output of oc version looks weird.

**Comment by shellyyang1989** [comment]
Logged into the upgrade pod and ran the upgrade:
```
sh-4.2$ export PATH=/tmp/client:$PATH
sh-4.2$ echo $PATH
/tmp/client:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
sh-4.2$ which oc
/tmp/client/oc
sh-4.2$ oc version
Client Version: openshift-clients-4.5.0-202006231303.p0-21-g297a4ac42
Server Version: 4.4.33
Kubernetes Version: v1.17.1+5ef953f
sh-4.2$ oc adm upgrade --to-image=registry.build03.ci.openshift.org/ci-op-lji8g5mn/release@sha256:c67fe644d1c06e6d7694e648a40199cb06e25e1c3cfd5cd4fdac87fd696d2297 --allow-explicit-upgrade --force=false
warning: The requested upgrade image is not one of the available updates.  You have used --allow-explicit-upgrade to the update to proceed anyway
Updating to release image registry.build03.ci.openshift.org/ci-op-lji8g5mn/release@sha256:c67fe644d1c06e6d7694e648a40199cb06e25e1c3cfd5cd4fdac87fd696d2297

sh-4.2$ oc adm upgrade
info: An upgrade is in progress. Working towards 4.5.41: 15% complete

warning: Cannot display available updates:
  Reason: NoChannel
  Message: The update channel has not been configured.
```
It looks good. It's not clear to me why it tells `Error: unknown flag: --allow-explicit-upgrade`.

**Comment by jianlinliu** [comment]
> It looks good. It's not clear to me why it tells Error: unknown flag: --allow-explicit-upgrade

Maybe need to print out some debug info in the codes:
```
echo $PATH
which oc
oc version
```

**Comment by shellyyang1989** [comment]
> > It looks good. It's not clear to me why it tells Error: unknown flag: --allow-explicit-upgrade
> 
> Maybe need to print out some debug info in the codes:
> 
> ```
> echo $PATH
> which oc
> oc version
> ```

They are printed out:
```
/tmp/client/oc
Client Version: version.Info{Major:"4", Minor:"0+", GitVersion:"v4.0.0-alpha.0+1c938ec-1967", GitCommit:"1c938ec", GitTreeState:"", BuildDate:"2019-04-10T10:43:26Z", GoVersion:"", Compiler:"", Platform:""}
```

**Comment by jianlinliu** [comment]
> They are printed out:
> 
> ```
> /tmp/client/oc
> Client Version: version.Info{Major:"4", Minor:"0+", GitVersion:"v4.0.0-alpha.0+1c938ec-1967", GitCommit:"1c938ec", GitTreeState:"", BuildDate:"2019-04-10T10:43:26Z", GoVersion:"", Compiler:"", Platform:""}
> ```

The oc version is too old.
Per the log:
```
oc adm release extract -a /var/run/secrets/ci.openshift.io/cluster-profile/pull-secret --command=oc --to=/tmp/client-2 registry.build03.ci.openshift.org/ci-op-wjvj1jin/release@sha256:c67fe644d1c06e6d7694e648a40199cb06e25e1c3cfd5cd4fdac87fd696d2297
```
The oc was extracted to `/tmp/client-2`, but not `/tmp/client`

**Comment by jiajliu** [comment]
> The oc was extracted to `/tmp/client-2`, but not `/tmp/client`
It was moved to `/tmp/client` from `/tmp/client-2` later. Could we print out the oc version info when it's in `/tmp/client-2`, just after extract and before mv to new dir?

**Comment by shellyyang1989** [comment]
Thanks all for your suggestions! I suspect it's the issue of old ansible image. It can proceed w/o the issue after I change the image to 4.12. https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/33715/rehearse-33715-periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-nightly-4.12-upgrade-from-stable-4.3-gcp-ipi/1590590114088620032

**Pull request closed without merging**